rust/src/vector.rs

```diff
@@ -0,0 +1,398 @@
+        let n = self.len();
+        let mut ptr = std::ptr::null();
+        unsafe {
+            bind_ceed::CeedVectorGetArrayRead(self.ptr, mtype as bind_ceed::CeedMemType, &mut ptr)
+        };
+        let ptr_slice: &[f64] = unsafe { slice::from_raw_parts(ptr, n) };
+        ndarray::aview1(ptr_slice)
+    }
+
+    /// Restore read access to a CeedVector via the specified memory type
+    ///
+    /// # arguments
+    ///
+    /// * 'array' - Array of vector data
+    ///
+    /// ```
+    /// # let ceed = ceed::Ceed::default_init();
+    /// let vec = ceed.vector(10);
+    /// let array = vec.get_array_read(ceed::MemType::Host);
+    /// vec.restore_array_read(array);
+    /// ```
+    pub fn restore_array_read(&self, array: ndarray::ArrayView1<f64>) {
+        let mut ptr = std::ptr::null();
+        unsafe { bind_ceed::CeedVectorRestoreArrayRead(self.ptr, &mut ptr) };
+        drop(array);
+    }
+
+    /// Create an immutable view
+    ///
+    /// ```
+    /// # let ceed= ceed::Ceed::default_init();
+    /// let vec = ceed.vector_from_slice(&[10., 11., 12., 13.]);
+    /// let v = vec.view();
+    /// assert_eq!(v[0..2], [10., 11.]);
+    /// // It is valid to have multiple immutable views
+    /// let w = vec.view();
+    /// assert_eq!(v[1..], w[1..]);
+    /// ```
+    pub fn view(&self) -> VectorView {
+        VectorView::new(self)
+    }
+
+    /// Return the norm of a CeedVector
+    ///
+    /// # arguments
+    ///
+    /// * 'ntype' - Norm type CEED_NORM_1, CEED_NORM_2, or CEED_NORM_MAX
+    ///
+    /// ```
+    /// # let ceed = ceed::Ceed::default_init();
+    /// let mut x = ceed.vector(10);
+    /// x.set_value(42.0);
+    /// let norm = x.norm(ceed::NormType::Max);
+    /// assert_eq!(norm, 42.0)
+    /// ```
+    pub fn norm(&self, ntype: crate::NormType) -> f64 {
+        let mut res: f64 = 0.0;
+        unsafe { bind_ceed::CeedVectorNorm(self.ptr, ntype as bind_ceed::CeedNormType, &mut res) };
+        res
+    }
+}
+
+// -----------------------------------------------------------------------------
+
+/// A (host) view of a Vector with Deref to slice.  We can't make
+/// Vector itself Deref to slice because we can't handle the drop to
+/// call bind_ceed::CeedVectorRestoreArrayRead().
+#[derive(Debug)]
+pub struct VectorView<'a> {
+    vec: &'a Vector<'a>,
+    array: Rc<*const f64>,
+}
+
+impl<'a> VectorView<'a> {
+    fn new(vec: &'a Vector) -> Self {
+        if let Some(array) = vec.array_weak.borrow().upgrade() {
+            return Self {
+                vec: vec,
+                array: Rc::clone(&array),
+            }
+        }
+        let mut ptr = std::ptr::null();
+        unsafe {
+            bind_ceed::CeedVectorGetArrayRead(vec.ptr, crate::MemType::Host as bind_ceed::CeedMemType, &mut ptr);
+        }
+        let array = std::rc::Rc::new(ptr);
+        vec.array_weak.replace(Rc::downgrade(&array));
+        Self {
+            vec: vec,
+            array: array,
+        }
+    }
+}
```

Comment:
An alternate way to do this would be to always call `CeedVectorGetArrayRead` when creating a view. I used reference counting on the Rust side (`Vector` holds a `Weak` reference while each `VectorView` holds a strong (`Rc`)) because it seems to let the compiler be a little stronger about enforcing the mutable reference rules. I'm not attached to it, however.